sources/RuleRunner.cpp

```diff
@@ -1329,6 +1330,96 @@ ActionHasTarget(int8 type)
 }
 
 
+static bool
+getDirectoryPath(BString& str, const entry_ref& ref)
+{
+	BEntry entry(&ref);
+	if (entry.InitCheck() != B_OK)
+		return false;
+
+	if (!entry.IsDirectory() && entry.GetParent(&entry) != B_OK)
+		return false;
+
+	BPath path(&entry);
+	if (path.InitCheck() != B_OK)
+		return false;
+
+	str = path.Path();
+	return true;
+}
+
+
+bool
+SetTextForType(BString& text, int8 type, const entry_ref& ref, bool isTest)
+{
+	if (isTest)
+		switch (type) {
+			case TEST_TYPE:
+			{
+				BNode node(&ref);
+				if (node.InitCheck() != B_OK)
+					return false;
+
+				BNodeInfo nodeInfo(&node);
+				if (nodeInfo.InitCheck() != B_OK)
+					return false;
+
+				char mimeType[B_MIME_TYPE_LENGTH];
+				if (nodeInfo.GetType(mimeType) != B_OK)
+					return false;
+
+				text = mimeType;
+				return true;
+			}
+			case TEST_NAME:
+				text = ref.name;
+				return true;
+			case TEST_SIZE:
+			{
+				BEntry entry(&ref);
+				if (entry.InitCheck() != B_OK || !entry.IsFile())
+					return false;
+
+				off_t size;
+				if (entry.GetSize(&size) != B_OK)
+					return false;
+
+				text = "";
+				text << size;
+				return true;
+			}
+			case TEST_LOCATION:
+				return getDirectoryPath(text, ref);
+			default:
+				return false;
+		}
+	else
+		switch (type) {
+			case ACTION_MOVE:
+			case ACTION_COPY:
+				return getDirectoryPath(text, ref);
+			case ACTION_RENAME:
+				text = ref.name;
+				return true;
+			case ACTION_ARCHIVE:
+			{
+				BEntry entry(&ref);
+				if (entry.InitCheck() != B_OK || !entry.IsFile())
+					return false;
+
+				BPath path(&ref);
+				if (path.InitCheck() != B_OK)
+					return false;
+
+				text = path.Path();
+				return true;
```

Comment:
There's another case you could handle for adding to an archive: If the user drops a folder instead of a file, you  could accept that and add a "/Archive.zip". The name "Archive.zip" should be localized.
If there's already an Archive.zip in the dropped folder, the new files will be added. Expected behaviour? You could add an increased number after the archive name if it already exists, but I guess that complicates it a bit...